lib/google/gax/bundling.rb

```diff
@@ -0,0 +1,445 @@
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @return [Event] an Event that can be used to wait on the response.
+      def extend(elts)
+        elts = [*elts]
+        @inputs.push(elts)
+        event = event_for(elts)
+        @events.push(event)
+        event
+      end
+
+      # Creates an Event that is set when the bundle with elts is sent.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @return [Event] an Event that can be used to wait on the response.
+      def event_for(elts)
+        event = Event.new
+        event.canceller = canceller_for(elts, event)
+        event
+      end
+
+      # Creates a cancellation proc that removes elts.
+      #
+      # The returned proc returns true if all elements were successfully removed
+      # from @inputs and  @events.
+      #
+      # @param elts [Array<Object>] an array of elements that can be appended
+      #     to the tasks bundle_field.
+      # @param [Event] an Event that can be used to wait on the response.
+      # @return [Proc] the canceller that when called removes the elts
+      #     and events.
+      def canceller_for(elts, event)
+        proc do
+          event_index = @events.find_index(event) || -1
+          in_index = @inputs.find_index(elts) || -1
+          @events.delete_at(event_index) unless event_index == -1
+          @inputs.delete_at(in_index) unless in_index == -1
+          if event_index == -1 || in_index == -1
+            false
+          else
+            true
+          end
+        end
+      end
+
+      private :run_with_no_subresponse,
+              :run_with_subresponses,
+              :event_for,
+              :canceller_for
+    end
+
+    # Organizes bundling for an api service that requires it.
+    class Executor
+      # @param options [BundleOptions]configures strategy this instance
+      #     uses when executing bundled functions.
+      # @param timer [Timer] the timer is used to handle the functionality of
+      #     timing threads.
+      def initialize(options, timer: Timer.new)
+        @options = options
+        @tasks = {}
+        @timer = timer
+
+        # Use a Monitor in order to have the mutex behave like a reentrant lock.
+        @tasks_lock = Monitor.new
+      end
+
+      # Schedules bundle_desc of bundling_request as part of
+      # bundle id.
+      #
+      # @param api_call [Proc] used to make an api call when the task is run.
+      # @param bundle_id [String] the id of this bundle.
+      # @param bundle_desc [BundleDescriptor] describes the structure of the
+      #     bundled call.
+      # @param bundling_request [Object] the request to pass as the arg to
+      #     the api_call.
+      # @return [Event] an Event that can be used to wait on the response.
+      #
+      # TODO: Implement byte and element count limits.
```

Comment:
nit: it's probably better to move this TODO inside of the method. The comment here will appear as a part of public documentation.